### ISSUE TITLE
Heap heuristics: cant detect libc version

### ISSUE THREAD

**Issue by disconnect3d** [pwndbg/pwndbg#1065]
This can be reproduced with the challenge from: https://github.com/perfectblue/ctf-writeups/tree/master/2019/insomnihack-teaser-2019/nyanc/challenge

To run this binary one has to do:
```
patchelf --set-interpreter ./ld-linux-aarch64.so.1 ./nyanc 
patchelf --replace-needed libc.so.6 ./libc.so.6 ./nyanc 
```
And then do `qemu-aarch64 -g 1234 ./nyanc` in one console and `gdb-multiarch` and `target remote :1234` in another one.

The error occurring is:
```
pwndbg> b main
pwndbg> continue
pwndbg> heap
'heap': Iteratively print chunks on a heap, default to the current thread's active heap.
Traceback (most recent call last):
  File "/home/user/pwndbg/pwndbg/commands/__init__.py", line 129, in __call__
    return self.function(*args, **kwargs)
  File "/home/user/pwndbg/pwndbg/commands/__init__.py", line 223, in _OnlyWhenRunning
    return function(*a, **kw)
  File "/home/user/pwndbg/pwndbg/commands/__init__.py", line 261, in _OnlyWithResolvedHeapSyms
    return function(*a, **kw)
  File "/home/user/pwndbg/pwndbg/commands/__init__.py", line 240, in _OnlyWhenHeapIsInitialized
    if pwndbg.heap.current.is_initialized():
  File "/home/user/pwndbg/pwndbg/heap/ptmalloc.py", line 1392, in is_initialized
    return any("heap]" in x.objfile for x in pwndbg.vmmap.get()) or self.main_arena['top'] != 0
  File "/home/user/pwndbg/pwndbg/heap/ptmalloc.py", line 685, in main_arena
    if pwndbg.glibc.get_version() < (2, 34) and pwndbg.arch.current != "arm":
TypeError: '<' not supported between instances of 'NoneType' and 'tuple'
```

How can we deal with it? 🤐  There are **no** symbols here as libc has them stripped (also not sure if GDB can fetch them if they were there...).

**Comment by lebr0nli** [comment]
> How can we deal with it? 🤐 There are **no** symbols here as libc has them stripped

If there are no symbols, is it possible to recover them without ida or ghidra?
If that is possible, I think we could use more heuristics to find some normal symbols we needed then we can still use the original heuristic to resolve the heap?

> (also not sure if GDB can fetch them if they were there...)

Hmm, I think it should work if they were there? 
By the way, is this libc really have **no** symbols? Because I can see symbols(not debug symbol) in ida and GDB.
I think here failed just because pwndbg's vmmap didn't fully support QEMU now, but `pwndbg.glibc.get_version()` needs it, and many other features in our heuristic need it, so error occurred.

You can temporary fixed it by comment the QEMU check in `pwndbg/vmmap.py` if you are using your patch version of QEMU:
(I didn't take a look inside the vmmap implementation, so this fix is still buggy, but it should be enough to test our heuristic.)
```diff
     # If we debug remotely a qemu-user or qemu-system target,
     # there is no point of hitting things further
-    if pwndbg.qemu.is_qemu():
-        return tuple()
+    # if pwndbg.qemu.is_qemu():
+    #     return tuple()
```
Then here is what I did:
```
pwndbg> set solib-search-path /ctf/work/challenge/lib
pwndbg> set solib-absolute-prefix /ctf/work/challenge/lib
pwndbg> target remote :1234
pwndbg> b main
pwndbg> c
pwndbg> arena
{
  mutex = 0,
  flags = 0,
  have_fastchunks = 0,
  fastbinsY = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0},
  top = 0x0,
  last_remainder = 0x0,
  bins = {0x0 <repeats 254 times>},
  binmap = {0, 0, 0, 0},
  next = 0x5502999a70,
  next_free = 0x0,
  attached_threads = 1,
  system_mem = 0,
  max_system_mem = 0,
}
pwndbg> mp
{
  trim_threshold = 131072,
  top_pad = 131072,
  mmap_threshold = 131072,
  arena_test = 8,
  arena_max = 0,
  n_mmaps = 0,
  n_mmaps_max = 65536,
  max_n_mmaps = 0,
  no_dyn_threshold = 0,
  mmapped_mem = 0,
  max_mmapped_mem = 0,
  sbrk_base = 0x0,
  tcache_bins = 64,
  tcache_max_bytes = 1032,
  tcache_count = 7,
  tcache_unsorted_limit = 0,
}
```
As you can see, we find correct `main_arena` and `mp_`

But somehow our heuristic seems broken for finding `tcache`, `thread_arena`, and `global_max_fast` with this libc 😢.

----

I think I almost fixed the bugs for these symbols, I'll open a new PR for this and for some issues you mentioned in previous PR!

**Comment by CptGibbon** [comment]
Thank you for your continued work on this ❤️

**Comment by disconnect3d** [comment]
@lebr0nli what's the state of this after #1075?

**Comment by lebr0nli** [comment]
> @lebr0nli what's the state of this after #1075?

@disconnect3d, I think for the error you saw at the beginning, it's fixed, but not because of  #1075, I guess it's actually because of one of the recent updates about the `vmmap` command (I'm not sure which commit).
The bug fixed in #1075 is that we assume all the functions of `vmmap` work great and GDB can find libc's symbol we used in the heuristics, then this PR can let heap command work without any issue at least for this libc.

And for now, if we using the stable version of QEMU, the output of `vmmap` is still a little bit weird (it shows some `.xxxx` pages, is this intended?). And also, it seems that we can't get some address of the libc symbol with the stable version of QEMU, and this cause many of the heuristics don't work, because we need to know the address of `_IO_list_all` and `__malloc_hook` in some functions.

> As you can see in the picture,  pwndbg can get the version of GLIBC now, but still can't find the correct mapping, and GDB with QEMU can't detect some symbols.
> <img width="851" alt="截圖 2022-09-15 上午9 52 11" src="https://user-images.githubusercontent.com/61896187/190294691-4cb0f6c8-a991-4710-8771-a96941f06a20.png">


And somehow, the new version of the `vmmap` command also does not work for your patch version of the QEMU, it returns a wrong page for `_IO_list_all`, and this cause many of the heuristics don't work too.

> As you can see in the picture, `vmmap` for your patch version of the QEMU have some weird `rwxp` pages and a weird `.comment` page.
> <img width="1043" alt="截圖 2022-09-15 上午2 27 58" src="https://user-images.githubusercontent.com/61896187/190288864-cae7a4fe-e04e-41a3-b0f1-181e962a3b18.png">

**Edit**: weird, this result only happened after we `continue` after hitting the breakpoint of `main` in this case...

----

But anyway, I think no matter what QEMU we are using, I guess it's possible to find where are the other symbols in libc without using `_IO_list_all`, and I'll look into it and see what I can do.

**Comment by lebr0nli** [comment]
@disconnect3d, btw you can use this branch: https://github.com/lebr0nli/pwndbg/tree/qemu-vmmap-locally to check whether heap heuristics works or not for this case more easily.

After `set resolve-qemu-vmmap-locally on` and using your patched QEMU in the same device, it should work without the above issues and output the correct mapping by the `vmmap` command for easier debugging.
(btw if you think we should create this feature, I can also open a PR for this :) )

**Comment by disconnect3d** [comment]
@lebr0nli The weird '.<name>' display in vmmap was caused by us fetching vmmap info the same way as we do for coredump files. This should be fixed in https://github.com/pwndbg/pwndbg/pull/1149

Regarding your idea with `set resolve-qemu-vmmap-locally on` it is nice, but maybe we should go along with what GEF does? If I recall correctly, for qemu-user emulation they were reading `/proc/$pid/maps` and filtering out memory pages from those of the qemu binary. Maybe this is the way to go?

We should do something like this and we should do it only for qemu-user emulation and only if we are sure we are connected to the local QEMU process, so we need to ensure that the /proc/$pid/maps is really the qemu process? Or maybe there is a better way, not sure.

Ideally, eventually we want to do something like `remote get /proc/$pid/maps` and just ask QEMU gdbstub to give us the debugee memory maps file like this. However, this is not implemented in QEMU. I have started some work around that in https://lore.kernel.org/all/20220221030910.3203063-1-dominik.b.czarnota@gmail.com/ but I need to come back to it and e.g. add tests.

**Comment by lebr0nli** [comment]
@disconnect3d, seems like with `patchelf`, the heap heuristics will not work for this case because gdbstub won’t detect the libc file, idk why :/ (Could it be a bug of qemu?)

> Somehow it will show:
>```
>warning: Probes-based dynamic linker interface failed.
>Reverting to original interface.
>```

But if we run the original `nyanc` without `patchelf` with: `qemu-aarch64 -L /work/challenge/ -g 1234 /work/challenge/nyanc`, then seems heap heuristics work normally after the recent updates for it.
The only inconvenient part is that we need to manually `set glibc 2.27` since vmmap is not work. (This should be fixed if we apply the enhancement in #1656)

**Comment by disconnect3d** [comment]
> then seems heap heuristics work normally after the recent updates for it.

Since this seems resolved, I am going to close this.
